src/architecture/merkledb.md

```diff
@@ -0,0 +1,346 @@
+# Exonum MerkleDB
+
+Exonum MerkleDB is a persistent storage implementation based on RocksDB
+which provides APIs to work with merklized data structures.
+
+MerkleDB is an object database. Objects represent the highest abstraction level
+for data storage.
+
+- All objects fall into the following groups: collections, blobs, and special
+  objects
+- Collections can have unlimited nesting
+- Objects of all groups can have hashed or non-hashed variants
+
+There are also root objects that don't have parents and have UTF-8 identifiers,
+for example "block", "state".
+
+!!! Warning
+Currently, only root objects with blob elements are supported.
+
+1. [Exonum object types](#object-types) lists supported types of
+   data storage collections.
+2. [Low-level storage](#low-level-storage) explains how object are persisted
+   using RocksDB
+3. [View layer](#view-layer) describes the wrapper over the DB engine
+   that ensures atomicity of blocks and transactions
+4. [Indexing](#indexing) gives an insight how indices over structured data
+   can be built in Exonum
+
+## Object Types
+
+Root object performs the same role as in relational database
+management systems (RDBMSs). However, unlike RDBMS tables,
+all MerkleDB objects internally are implemented as wrappers around key-value
+stores.
+Both keys and values in the wrapped stores are persisted as byte sequences.
+MerkleDB does not natively support operations (matching, grouping, sorting, etc.)
+over separate value fields, as it is the case with other key-value storages.
+
+### Key Sorting and Iterators
+
+MerkleDB objects implement iterators over stored items (or keys, values, and
+key-value
+pairs in the case of maps). Such
+iterators use key ordering of the underlying key-value storage to determine
+the iteration order.
+Namely, keys are lexicographically ordered over their binary serializations;
+this ordering coincides with that used in
+[RocksDB](#low-level-storage).
+
+### MapIndex
+
+[`MapIndex`][map-index] implements a key-value store aka a map. It has
+the following functionality:
+
+- Get, set and remove value by key
+- Check if a specific key is present in the map
+- Iterate over the key-value pairs in the lexicographic key order
+- Iterate over keys in the lexicographic key order
+- Iterate over values in the lexicographic key order
+- Clear the map (i.e., remove all stored key-value pairs)
```

Comment:
We've added `.len()` and `.is_empty()` methods to the [Proof]MapIndex. Should we add this info here ?